File: scripts/generate-image.py
#!/usr/bin/env python3
import argparse
import json
import os
import sys
import urllib.request
from pathlib import Path
from datetime import datetime

try:
    from PIL import Image
    HAS_PIL = True
except ImportError:
    HAS_PIL = False

try:
    from rembg import remove as rembg_remove
    HAS_REMBG = True
except ImportError:
    HAS_REMBG = False

MODELS = {
    "gpt-image-1-mini": {
        "t2i": "fal-ai/gpt-image-1-mini",
        "i2i": "fal-ai/gpt-image-1-mini/edit",
        "image_field": "image_urls",
        "use_image_size": True,
        "default_params": {"quality": "low", "background": "auto"},
    },
    "grok-imagine": {
        "t2i": "xai/grok-imagine-image",
        "i2i": "xai/grok-imagine-image/edit",
        "image_field": "image_url",
        "use_aspect_ratio": True,
    },
    "gemini-25-flash-image": {
        "t2i": "fal-ai/gemini-25-flash-image",
        "i2i": "fal-ai/gemini-25-flash-image/edit",
        "image_field": "image_urls",
        "use_aspect_ratio": True,
        "default_params": {"safety_tolerance": "4"},
    },
    "gemini-3-pro": {
        "t2i": "fal-ai/gemini-3-pro-image-preview",
        "i2i": "fal-ai/gemini-3-pro-image-preview/edit",
        "image_field": "image_urls",
        "use_aspect_ratio": True,
        "default_params": {"safety_tolerance": "4", "resolution": "1K"},
    },
    "gpt-image-1.5": {
        "t2i": "fal-ai/gpt-image-1.5",
        "i2i": "fal-ai/gpt-image-1.5/edit",
        "image_field": "image_urls",
        "use_image_size": True,
        "default_params": {"quality": "medium", "background": "auto"},
    },
}

ASPECT_TO_IMAGE_SIZE = {
    "1:1": "1024x1024",
    "4:3": "1536x1024",
    "3:2": "1536x1024",
    "16:9": "1536x1024",
    "21:9": "1536x1024",
    "3:4": "1024x1536",
    "2:3": "1024x1536",
    "9:16": "1024x1536",
}

TAURI_SIZES = [32, 128, 256, 512]


def detect_output_dir(base_path: Path) -> Path:
    if (base_path / "src-tauri" / "icons").exists():
        return base_path / "src-tauri" / "icons"
    if (base_path / "public").exists():
        return base_path / "public" / "images" / "generated"
    if (base_path / "assets").exists():
        return base_path / "assets" / "images"
    return base_path / "generated-images"


def load_fal_key() -> str:
    key = os.environ.get("FAL_KEY") or os.environ.get("FAL_API_KEY")
    if not key:
        for env_file in [".env.local", ".env"]:
            p = Path(env_file)
            if p.exists():
                for line in p.read_text().splitlines():
                    if "=" in line and line.split("=", 1)[0].strip() in ("FAL_KEY", "FAL_API_KEY"):
                        key = line.split("=", 1)[1].strip().strip('"').strip("'")
                        break
            if key:
                break
    if not key:
        raise RuntimeError("FAL_KEY not found. Add FAL_KEY=... to .env.local")
    return key


def call_fal_api(endpoint: str, payload: dict, api_key: str) -> dict:
    data = json.dumps(payload).encode()
    req = urllib.request.Request(
        f"https://fal.run/{endpoint}",
        data=data,
        headers={"Authorization": f"Key {api_key}", "Content-Type": "application/json"},
        method="POST",
    )
    with urllib.request.urlopen(req, timeout=120) as resp:
        return json.loads(resp.read())


def download_image(url: str, dest: Path):
    dest.parent.mkdir(parents=True, exist_ok=True)
    urllib.request.urlretrieve(url, dest)


def build_payload(
    prompt: str,
    model_key: str,
    model_cfg: dict,
    image_url: str | None,
    aspect: str,
    quality: str | None,
    output_format: str,
) -> dict:
    payload: dict = {"prompt": prompt, "num_images": 1, "output_format": output_format}

    if model_cfg.get("use_image_size"):
        payload["image_size"] = ASPECT_TO_IMAGE_SIZE.get(aspect, "1024x1024")
    elif model_cfg.get("use_aspect_ratio"):
        payload["aspect_ratio"] = aspect

    if model_cfg.get("default_params"):
        payload.update(model_cfg["default_params"])

    if quality and "quality" in payload:
        payload["quality"] = quality

    if image_url:
        field = model_cfg["image_field"]
        payload[field] = [image_url] if field == "image_urls" else image_url

    return payload


def apply_post_process(image_path: Path, post_process: str, output_path: Path) -> Path:
    if post_process == "none":
        if image_path != output_path:
            image_path.rename(output_path)
        return output_path

    if post_process == "rembg":
        if not HAS_REMBG:
            raise RuntimeError("rembg not installed. Run: pip install rembg")
        result = rembg_remove(image_path.read_bytes())
        out = output_path.with_suffix(".png")
        out.write_bytes(result)
        return out

    if post_process.startswith("resize:"):
        if not HAS_PIL:
            raise RuntimeError("Pillow not installed. Run: pip install pillow")
        w, h = map(int, post_process.split(":")[1].split("x"))
        img = Image.open(image_path).resize((w, h), Image.LANCZOS)
        img.save(output_path)
        return output_path

    if post_process == "ico":
        if not HAS_PIL:
            raise RuntimeError("Pillow not installed. Run: pip install pillow")
        img = Image.open(image_path).convert("RGBA")
        ico_path = output_path.with_suffix(".ico")
        img.save(ico_path, format="ICO", sizes=[(16, 16), (32, 32), (48, 48), (64, 64)])
        img.resize((180, 180), Image.LANCZOS).save(output_path.parent / "apple-touch-icon.png")
        print(f"   + apple-touch-icon.png")
        return ico_path

    if post_process == "tauri-icons":
        if not HAS_PIL:
            raise RuntimeError("Pillow not installed. Run: pip install pillow")
        img = Image.open(image_path).convert("RGBA")
        for size in TAURI_SIZES:
            out = output_path.parent / f"{size}x{size}.png"
            img.resize((size, size), Image.LANCZOS).save(out)
            print(f"   + {out.name}")
        ico_path = output_path.parent / "icon.ico"
        img.resize((256, 256), Image.LANCZOS).save(
            ico_path, format="ICO", sizes=[(32, 32), (64, 64), (128, 128), (256, 256)]
        )
        print(f"   + icon.ico")
        return output_path.parent

    return image_path


def generate(args):
    api_key = load_fal_key()

    model_cfg = MODELS.get(args.model)
    if not model_cfg:
        raise ValueError(f"Unknown model: {args.model}. Available: {', '.join(MODELS.keys())}")

    use_edit = bool(args.image)
    endpoint = model_cfg["i2i"] if use_edit else model_cfg["t2i"]

    payload = build_payload(
        args.prompt, args.model, model_cfg, args.image, args.aspect, args.quality, args.format
    )

    print(f"\nGenerating image...")
    print(f"  Model   : {args.model} ({'edit' if use_edit else 'text-to-image'})")
    print(f"  Endpoint: {endpoint}")
    print(f"  Prompt  : {args.prompt[:80]}{'...' if len(args.prompt) > 80 else ''}")

    result = call_fal_api(endpoint, payload, api_key)
    images = result.get("images", [])
    if not images:
        raise RuntimeError(f"No images returned: {result}")

    image_url = images[0].get("url") or images[0]

    if args.output:
        output_path = Path(args.output)
    else:
        out_dir = detect_output_dir(Path.cwd())
        ts = datetime.now().strftime("%Y%m%d-%H%M%S")
        ext = "jpg" if args.format == "jpeg" else args.format
        output_path = out_dir / f"{args.type}-{ts}.{ext}"

    output_path.parent.mkdir(parents=True, exist_ok=True)

    raw_path = output_path.with_stem(output_path.stem + "-raw")
    print(f"  Downloading...")
    download_image(image_url, raw_path)

    final = apply_post_process(raw_path, args.post_process, output_path)

    if raw_path.exists() and raw_path != final:
        raw_path.unlink()

    print(f"  Saved: {final}")
    return final


def main():
    parser = argparse.ArgumentParser(description="Generate images via fal.ai")
    parser.add_argument("prompt", help="Image generation prompt")
    parser.add_argument(
        "--mode", choices=["t2i", "i2i"], default="t2i",
        help="t2i = text-to-image, i2i = image-to-image (edit)"
    )
    parser.add_argument(
        "--type",
        choices=["layout", "hero", "icon", "favicon", "mascot", "background", "illustration"],
        default="layout",
    )
    parser.add_argument("--model", default="gemini-25-flash-image", choices=list(MODELS.keys()))
    parser.add_argument("--aspect", default="16:9", help="Aspect ratio, e.g. 16:9, 1:1, 9:16")
    parser.add_argument("--image", help="Reference image path or URL (required for i2i)")
    parser.add_argument("--post-process", default="none", dest="post_process",
                        help="none | rembg | resize:WxH | ico | tauri-icons")
    parser.add_argument("--output", help="Output file path (auto-detected if omitted)")
    parser.add_argument("--quality", choices=["low", "medium", "high"])
    parser.add_argument("--format", default="png", choices=["png", "jpeg", "webp"])

    args = parser.parse_args()

    if args.mode == "i2i" and not args.image:
        parser.error("--image is required for i2i mode")

    generate(args)


if __name__ == "__main__":
    main()

File: scripts/generate-image.mjs
#!/usr/bin/env node
/**
 * generate-image.mjs — Geração de imagem via FAL.AI (zero-dep Node).
 *
 * Lê models/image-models.json como fonte única de verdade.
 * Funciona em qualquer máquina com Node 18+ e FAL_AI_API_KEY (não depende
 * de generate.py em path privado).
 *
 * CLI:
 *   node scripts/generate-image.mjs --prompt "..." --out cover.jpg
 *   node scripts/generate-image.mjs --prompt "..." --model gemini-25-flash --ref photo.jpg --out edit.png
 *   node scripts/generate-image.mjs --list
 *
 * Lib:
 *   import { generateImage, listModels } from "scripts/generate-image.mjs"
 *
 * Default rule (sem --model explícito):
 *   - Sem --ref       → grok-imagine (text-to-image, $0.020)
 *   - Com --ref       → gemini-25-flash (edit, $0.039)
 *
 * Auth:
 *   FAL_AI_API_KEY (fallback: FAL_KEY, FAL_API_KEY)
 *
 * Policy: skills/17-image-generator/SKILL.md
 */

import { readFileSync, writeFileSync, existsSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { dirname, join, resolve, basename } from "node:path";

// ──────────────────────────────────────────────────────────────
// Models config (single source of truth)
// ──────────────────────────────────────────────────────────────

const __dirname = dirname(fileURLToPath(import.meta.url));
const MODELS_PATH = join(__dirname, "..", "models", "image-models.json");

function loadModels() {
  if (!existsSync(MODELS_PATH)) {
    throw new Error(`Models config missing: ${MODELS_PATH}`);
  }
  return JSON.parse(readFileSync(MODELS_PATH, "utf8"));
}

// ──────────────────────────────────────────────────────────────
// Auth
// ──────────────────────────────────────────────────────────────

function getApiKey() {
  const key =
    process.env.FAL_AI_API_KEY ??
    process.env.FAL_KEY ??
    process.env.FAL_API_KEY;
  if (!key) {
    throw new Error(
      "FAL_AI_API_KEY não definida. Setup: https://fal.ai/dashboard/keys"
    );
  }
  return key;
}

// ──────────────────────────────────────────────────────────────
// Default model resolution (the rule)
// ──────────────────────────────────────────────────────────────

function resolveModel(opts, config) {
  if (opts.model) {
    if (!config.models[opts.model]) {
      throw new Error(
        `Unknown model: ${opts.model}. Available: ${Object.keys(config.models).join(", ")}`
      );
    }
    return opts.model;
  }
  const hasRef = (opts.referenceImages?.length ?? 0) > 0;
  const presetName = hasRef ? config.default_rule.if_has_reference_images : config.default_rule.else;
  const preset = config.presets[presetName];
  if (!preset) throw new Error(`Preset não encontrado: ${presetName}`);
  return preset.model;
}

// ──────────────────────────────────────────────────────────────
// FAL.AI REST client (submit-then-poll, sem SDK)
// ──────────────────────────────────────────────────────────────

async function falCall(endpoint, body, queueBaseUrl) {
  const apiKey = getApiKey();

  const submitRes = await fetch(`${queueBaseUrl}/${endpoint}`, {
    method: "POST",
    headers: {
      Authorization: `Key ${apiKey}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify(body),
  });

  if (!submitRes.ok) {
    const text = await submitRes.text();
    throw new Error(`FAL submit failed (${submitRes.status}): ${text}`);
  }

  const submitData = await submitRes.json();
  const { status_url, response_url } = submitData;

  // Poll status
  const start = Date.now();
  const timeoutMs = 90_000;
  while (Date.now() - start < timeoutMs) {
    await new Promise(r => setTimeout(r, 1500));
    const statusRes = await fetch(status_url, {
      headers: { Authorization: `Key ${apiKey}` },
    });
    if (!statusRes.ok) continue;
    const status = await statusRes.json();
    if (status.status === "COMPLETED") {
      const resultRes = await fetch(response_url, {
        headers: { Authorization: `Key ${apiKey}` },
      });
      if (!resultRes.ok) {
        const text = await resultRes.text();
        throw new Error(`FAL result fetch failed (${resultRes.status}): ${text}`);
      }
      return await resultRes.json();
    }
    if (status.status === "FAILED" || status.status === "ERROR") {
      throw new Error(`FAL job failed: ${JSON.stringify(status)}`);
    }
  }
  throw new Error(`FAL timeout after ${timeoutMs}ms`);
}

// ──────────────────────────────────────────────────────────────
// API pública (importável)
// ──────────────────────────────────────────────────────────────

export function listModels() {
  const config = loadModels();
  return Object.entries(config.models).map(([id, m]) => ({
    id,
    name: m.name,
    vendor: m.vendor,
    tier: m.tier,
    t2i_price: m.pricing.t2i_usd_per_image,
    edit_price: m.pricing.edit_usd_per_image,
    use_cases: m.use_cases,
  }));
}

/**
 * @param {Object} opts
 * @param {string} opts.prompt — texto da imagem desejada
 * @param {string[]} [opts.referenceImages] — URLs/data URLs pra edit
 * @param {string} [opts.model] — força model específico (override do default rule)
 * @param {string} [opts.aspectRatio] — ex: "16:9", "1:1"
 * @param {number} [opts.numImages=1]
 * @param {string} [opts.outputFormat] — jpeg | png | webp
 * @param {string} [opts.outPath] — se setado, baixa primeira imagem pra esse path
 */
export async function generateImage(opts) {
  if (!opts.prompt) throw new Error("prompt é obrigatório");

  const config = loadModels();
  const modelId = resolveModel(opts, config);
  const model = config.models[modelId];
  const hasRef = (opts.referenceImages?.length ?? 0) > 0;

  if (hasRef && !model.supports.edit) {
    throw new Error(`Model ${modelId} não suporta edit. Use grok-imagine ou gemini-25-flash.`);
  }

  const endpoint = hasRef ? model.endpoints.edit : model.endpoints.t2i;

  const body = {
    prompt: opts.prompt,
    num_images: opts.numImages ?? model.default_params?.num_images ?? 1,
    output_format: opts.outputFormat ?? model.default_params?.output_format ?? "jpeg",
  };

  // Aspect ratio vs image_size — depende do model
  if (model.supports.aspect_ratios && opts.aspectRatio) {
    body.aspect_ratio = opts.aspectRatio;
  } else if (model.default_params?.aspect_ratio) {
    body.aspect_ratio = model.default_params.aspect_ratio;
  }
  if (model.supports.image_sizes && opts.imageSize) {
    body.image_size = opts.imageSize;
  } else if (model.default_params?.image_size && !body.aspect_ratio) {
    body.image_size = model.default_params.image_size;
  }
  if (model.default_params?.quality) body.quality = model.default_params.quality;

  if (hasRef) {
    body.image_urls = opts.referenceImages;
  }

  const start = Date.now();
  const result = await falCall(endpoint, body, config.queue_base_url);
  const durationMs = Date.now() - start;

  const images = (result.images ?? []).map(img => ({
    url: img.url,
    width: img.width,
    height: img.height,
    contentType: img.content_type ?? `image/${body.output_format}`,
  }));

  const out = {
    images,
    model: modelId,
    estimatedCostUsd: (hasRef ? model.pricing.edit_usd_per_image : model.pricing.t2i_usd_per_image) * images.length,
    durationMs,
  };

  // Download primeira imagem se outPath setado
  if (opts.outPath && images[0]?.url) {
    const res = await fetch(images[0].url);
    const buf = Buffer.from(await res.arrayBuffer());
    writeFileSync(opts.outPath, buf);
    out.savedTo = opts.outPath;
  }

  return out;
}

// ──────────────────────────────────────────────────────────────
// CLI
// ──────────────────────────────────────────────────────────────

function parseArgs(argv) {
  const args = { numImages: 1 };
  for (let i = 2; i < argv.length; i++) {
    const a = argv[i];
    switch (a) {
      case "--prompt":      args.prompt = argv[++i]; break;
      case "--model":       args.model = argv[++i]; break;
      case "--ref":
      case "--reference":   (args.referenceImages ??= []).push(argv[++i]); break;
      case "--aspect":      args.aspectRatio = argv[++i]; break;
      case "--size":        args.imageSize = argv[++i]; break;
      case "--n":
      case "--num-images":  args.numImages = parseInt(argv[++i], 10); break;
      case "--format":      args.outputFormat = argv[++i]; break;
      case "--out":         args.outPath = argv[++i]; break;
      case "--list":        args.list = true; break;
      case "--help":
      case "-h":            args.help = true; break;
    }
  }
  return args;
}

function printHelp() {
  const config = loadModels();
  console.log(`generate-image.mjs — FAL.AI image generation (zero-dep)

Usage:
  node scripts/generate-image.mjs --prompt "..." [options]

Required:
  --prompt <text>           Texto da imagem

Options:
  --model <id>              Força model específico (default: regra abaixo)
  --ref <url-or-path>       Imagem de referência (pode repetir). Ativa edit mode.
  --aspect <ratio>          Ex: "16:9", "1:1", "9:16"
  --size <size>             Ex: "1024x1024" (alguns models)
  --n <int>                 Quantas imagens (default: 1)
  --format <fmt>            jpeg | png | webp
  --out <path>              Baixa primeira imagem pro path local
  --list                    Lista models disponíveis
  --help                    Mostra este help

Default rule (sem --model):
  - Sem --ref  → ${config.default_text_to_image} (t2i)
  - Com --ref  → ${config.default_edit_image} (edit)

Auth: env var FAL_AI_API_KEY

Examples:
  node scripts/generate-image.mjs --prompt "minimalist cover, blue gradient" --out cover.jpg
  node scripts/generate-image.mjs --prompt "remove background" --ref photo.jpg --out clean.png
  node scripts/generate-image.mjs --prompt "OG card" --model gpt-image-1.5 --aspect 16:9 --out og.png
  node scripts/generate-image.mjs --list
`);
}

async function main() {
  const args = parseArgs(process.argv);
  if (args.help) { printHelp(); return; }
  if (args.list) {
    const models = listModels();
    console.log("Available models:");
    for (const m of models) {
      console.log(`  ${m.id.padEnd(20)} ${m.tier.padEnd(15)} $${m.t2i_price.toFixed(3)}/img — ${m.use_cases[0] ?? ""}`);
    }
    return;
  }
  if (!args.prompt) {
    console.error("Error: --prompt obrigatório. Use --help.");
    process.exit(1);
  }

  try {
    const result = await generateImage(args);
    console.log(`✓ ${result.images.length} image(s) generated via ${result.model}`);
    console.log(`  Cost: $${result.estimatedCostUsd.toFixed(4)}`);
    console.log(`  Time: ${result.durationMs}ms`);
    for (const img of result.images) {
      console.log(`  → ${img.url}`);
    }
    if (result.savedTo) console.log(`  Saved to: ${result.savedTo}`);
  } catch (err) {
    console.error("Error:", err.message);
    process.exit(2);
  }
}

// Run CLI se invocado diretamente
if (import.meta.url === `file://${process.argv[1].replace(/\\/g, "/")}` ||
    basename(process.argv[1] ?? "") === "generate-image.mjs") {
  main();
}
